### ISSUE TITLE
False positive error when inferring type of list expression

### ISSUE THREAD

**Issue by adsharma** [microsoft/pyright#1710]
**Describe the bug**

Details here: https://github.com/mwhittaker/quoracle/issues/4

It's not clear if the problem is with pyright or the way type annotations are done in expr.py in the repo above and reproduced in the screenshot section below.

**To Reproduce**

git clone https://github.com/mwhittaker/quoracle/
pyright

**Expected behavior**

Few type errors, but get 300+

**Screenshots or Code**
If applicable, add screenshots or the text of the code (surrounded by triple back ticks) to help explain your problem.
```python
class Expr(Generic[T]):
    def __add__(self, rhs: 'Expr[T]') -> 'Expr[T]':
     ...
     
class Node(Expr[T]):
   ...

a: Node[str] = ...
b: Node[str] = ...
c = a + b
```

**Comment by erictraut** [comment]
It would be useful if you could extract a specific self-contained example of where you believe that pyright is behaving incorrectly.

Pyright includes more complete checks than mypy with respect to operator overloads (e.g. `__add__` or `__mul__`). That may explain what you are seeing.

When I use the code you pasted above, I don't see any error reported.
```python
class Expr(Generic[T]):
    def __add__(self, rhs: 'Expr[T]') -> 'Expr[T]':
     ...
     
class Node(Expr[T]):
   ...

def func(a: Node[str], b: Node[str]):
    c = a + b
    reveal_type(c) # Expr[str]
```

**Comment by adsharma** [comment]
The most common error type is:

```
  18:22 - error: Operator "+" not supported for types "Node[Literal['a']]" and "Node[Literal['b']]"
```
in

```python
        a = Node('a')
        b = Node('b')
        c = Node('c')

        assert_equal(a+b+c, [{'a'}, {'b'}, {'c'}])
```

In other words, pyright doesn't understand the operator overloading. I'm not sure I understand why.

**Comment by erictraut** [comment]
Here's a complete, self-contained sample based on your code above. Pyright doesn't emit any errors for me. Are you seeing errors with this sample?

```python
from typing import Generic, Optional, TypeVar
import datetime

T = TypeVar("T")

class Expr(Generic[T]):
    def __add__(self, rhs: "Expr[T]") -> "Expr[T]":
        ...

class Node(Expr[T]):
    def __init__(
        self,
        x: T,
        capacity: Optional[float] = None,
        read_capacity: Optional[float] = None,
        write_capacity: Optional[float] = None,
        latency: datetime.timedelta = None,
    ) -> None:
        ...

a = Node("a")
b = Node("b")
c = Node("c")

d = a + b + c
```

**Comment by adsharma** [comment]
Here is the smallest self contained example I could produce:

https://paste.ubuntu.com/p/dJHKc3rnxJ/

I'm seeing just one error:

```
  269:26 - error: Argument of type "list[Node[Literal['a']] | Node[Literal['b']] | Node[Literal['c']]]" cannot be assigned to parameter "es" of type "List[Expr[T@choose]]" in function "choose"
    TypeVar "_T@list" is invariant
      Type "Node[Literal['a']] | Node[Literal['b']] | Node[Literal['c']]" cannot be assigned to type "Expr[T@choose]"
        "Node[Literal['a']]" is incompatible with "Expr[T@choose]"
        "Node[Literal['b']]" is incompatible with "Expr[T@choose]"
        "Node[Literal['c']]" is incompatible with "Expr[T@choose]" (reportGeneralTypeIssues)
```

The other error (`Operator "+" not supported`) is spurious. Please ignore.

**Comment by erictraut** [comment]
OK, thanks. Yes, this looks like a bug in the type checker — specifically with its inference logic for list expressions when the expected type is generic (in this case, `Expr[T]`). I'll investigate.

**Comment by erictraut** [comment]
Thanks for the bug report. I've introduced a fix for this particular issue, and that will be included in the next release of pyright. If you see others, please file separate bugs.

**Comment by adsharma** [comment]
Thanks. I'm still seeing `Operator "+" not supported for types...` when I run pyright on the git repo, just not a standalone repro of the issue. Hopefully it's a variant of the same bug and will go away when the fix is in.

**Comment by erictraut** [comment]
The `Operator "+"` issue is not related to this bug, so there's another issue you're seeing.

I cloned the quoracle repo, and I'm not able to repro any errors related to operator checking. Are you sure you have the latest version of the code checked out? If you're not able to repro it with a standalone piece of code, that leads me to believe it has to do with the code you have checked out or the Python environment or configuration settings you're using in that particular project.

**Comment by adsharma** [comment]
It was seen with `1.1.98`. Went away after I upgraded to `1.1.127`.

**Comment by erictraut** [comment]
This fix is included in pyright 1.1.128, which I just published. It will also be included in the next release of pylance.
